Maxwellian 3D speed fixes

lemmatum opened the pull request:
Supersedes #451 
Closes #392 

Our speed distributions are all wrong. The 1D Maxwellian speed distribution is actually the 3D distribution, so that is being renamed. I'm also adding the correct form for the 1D distribution and adding 2D distributions.

pep8speaks commented:
Hello @lemmatum! Thanks for submitting your pull request.



 - In the file [`plasmapy/physics/tests/test_distribution.py`](https://github.com/PlasmaPy/PlasmaPy/blob/b670da7ef0d8024ad39b67f222c39f8c89a593a5/plasmapy/physics/tests/test_distribution.py), following are the PEP8 issues :

> [Line 211:1](https://github.com/PlasmaPy/PlasmaPy/blob/b670da7ef0d8024ad39b67f222c39f8c89a593a5/plasmapy/physics/tests/test_distribution.py#L211): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '

PlasmaPy follows the [PEP8 style guide](https://www.python.org/dev/peps/pep-0008/), which may be checked locally using [pycodestyle](http://pycodestyle.pycqa.org/en/latest/intro.html)

lemmatum commented:
Here is an outline of what I'm doing
[maxwellian_speed_3D.pdf](https://github.com/PlasmaPy/PlasmaPy/files/1989773/maxwellian_speed_3D.pdf)

namurphy commented on plasmapy/physics/distribution.py:
```diff
@@ -143,6 +143,144 @@ def Maxwellian_1D(v,
+                           vTh=np.nan,
+                           units="units"):
+    r"""
+    pdf of velocity for a 2D Maxwellian distribution.
```
Perhaps spell out pdf as probability function, since not everyone will know the acronym.

dstansby commented on plasmapy/physics/distribution.py:
```diff
@@ -143,6 +143,144 @@ def Maxwellian_1D(v,
+    r"""
+    pdf of velocity for a 2D Maxwellian distribution.
+
+    Return the probability of finding a particle with velocity components
```
Technically this returns the probability density function, not the probability (to get a probability you have to multiply by `dv_x . dv_y`).

lemmatum replied:
done

dstansby commented on plasmapy/physics/distribution.py:
```diff
@@ -143,6 +143,144 @@ def Maxwellian_1D(v,
+        ``'D+'`` for deuterium, or ``'He-4 +1'`` for :math:`He_4^{+1}`
+        (singly ionized helium-4)), which defaults to electrons.
+
+    Vx_drift: ~astropy.units.Quantity, optional
```
Having lower case parameter names is a python convention, so I'd vote for `vx_drift` (but sometimes it's best to ignore that for e.g. magnetic field as `B`)

lemmatum replied:
Made them lowercase.

dstansby commented on plasmapy/physics/distribution.py:
```diff
@@ -143,6 +143,144 @@ def Maxwellian_1D(v,
         return distFunc
 
 
+def Maxwellian_velocity_2D(vx,
```
This doesn't seem to be appearing in the built docs: https://1504-46810954-gh.circle-artifacts.com/0/home/circleci/project/docs/_build/html/physics/distribution.html

lemmatum replied:
Do I have to manually add stuff in the docs?

lemmatum replied:
Looks like it's showing up now
https://1508-46810954-gh.circle-artifacts.com/0/home/circleci/project/docs/_build/html/physics/distribution.html?highlight=distribution#module-plasmapy.physics.distribution

dstansby commented on plasmapy/physics/distribution.py:
```diff
@@ -143,6 +143,144 @@ def Maxwellian_1D(v,
+                           vTh=np.nan,
+                           units="units"):
+    r"""
+    pdf of velocity for a 2D Maxwellian distribution.
```
Re-phrasing as "pdf of velocity for a Maxwellian distribution in 2D" would make it clearer that this is specifically for things in two dimensions.

lemmatum replied:
done

dstansby commented on plasmapy/physics/distribution.py:
```diff
@@ -143,6 +143,144 @@ def Maxwellian_1D(v,
+    Returns
+    -------
+    f : ~astropy.units.Quantity
+        Probability in Velocity^-1, normalized so that
```
This should be "Probability density"

lemmatum replied:
done throughout

dstansby commented on plasmapy/physics/distribution.py:
```diff
@@ -143,6 +143,144 @@ def Maxwellian_1D(v,
+        vTh = parameters.thermal_speed(T * u.K,
+                                       particle=particle,
+                                       method="most_probable").si.value
+    # accounting for thermal velocity in 3D
```
"3D" > "2D"

lemmatum replied:
fixed

dstansby commented on plasmapy/physics/distribution.py:
```diff
@@ -370,7 +508,129 @@ def Maxwellian_speed_1D(v,
+    .. math::
+
+       f = 2 \pi \vec{v} (\pi v_{Th}^2)^{-1} \exp(-(\vec{v} -
+       \vec{V}_{drift})^2 / v_{Th}^2)
```
I don't think this is correct; when the drift speed is not zero, it is non-trivial to integrate over angles to get a probability density function for speed; see my working here: https://www.dropbox.com/s/rsjb65nvl2v9kbh/max_2d.pdf?dl=0

lemmatum replied:
Doesn't the drift velocity just shift the position of distribution in velocity space? (thinking in Cartesian coordinates). Just thinking about it spatially/geometrically, this shouldn't affect the normalization coefficient.

Thinking about it numerically, if you want to get the normalization coefficient you can just make the substitution `v_new = v_old - v_drift`. Since `v_drift` is a constant we have `dv_new = dv_old` and we can just do the integration as we would in the case of `v_drift = 0`.

The one caveat here is that I think it modifies the velocity terms in the normalization coefficient. So for a 3D distribution we have `4 * pi * v ** 2` representing the spherical shell over which we are integrating. I think this shell gets shifted to be centered around the drifting Maxwellian, so it would be something like  `4 * pi * (v - v_drift) ** 2`.

dstansby replied:
It doesn't affect the normalisation coefficient, but it does effect the probability distribution function (pdf) of the radial velocity coordinate (ie speed) in a polar coordinate system - see my above derivation. Let me know if anything doesn't make sense, but I'm pretty sure that with a drift velocity the pdf of speed currently in the code is wrong.

lemmatum replied:
Yea I've taken a look at the derivation and I'm not convinced. What you argue in the derivation would imply an effect on the normalization coefficient (since the exponentials cannot be affected because they're exponentials), otherwise nothing is affected and the current implementation is correct.

I'll try to work this out on paper using change of variables.

dstansby replied:
My point is that doing the integral in my equation (6) isn't possible (or at least doesn't come out with the equation currently in the docstring) for v_drift != 0.

If v_drift = 0 the integral is possible, and results in the equation in the docstring (with v_drift = 0).

dstansby replied:
I've added a bit at the end with v_drift = 0, to show how it reduces to the expected equation which might help?

lemmatum replied:
So here is my derivation for the current implementation of the velocity and speed distribution including drift. This is using a change of variables to shift the velocity distribution to the origin, do the integration, and then shift back the normalized distribution(s).

![img_20180511_024812](https://user-images.githubusercontent.com/28945309/39915837-38aed98a-54c6-11e8-9f61-59c1d18cdbdf.jpg)

dstansby replied:
I don't think the final equation can be correct, as it goes negative for v > v_drift (and probability densities can't be negative!). I think the problem is that the final "shifting back" stage is not valid - shifting in velocity space is not the same as shifting in speed (radial coordinate) space.

dstansby replied:
Yeah, in that last step, you have to take into account the differential bit when changing coordinates, ie. 
![image](https://user-images.githubusercontent.com/6197628/39916752-5a7625e0-5503-11e8-9539-50fa60e4241b.png)
 and I think you have assumed that dv/dv' = 1, which isn't true.

lemmatum replied:
hmmm, yea it goes negative for `v < v_drift`... perhaps the normalization constant is just `v` then, since the size of the area of integration doesn't change due to drift offsetting the distribution in velocity space.

Good point about shifting in velocity vs in speed space. There must be a way to convert the shift in velocity space to one in speed space. I think that will involve an absolute value since `v = |\vec{v}|`... so maybe the `v - v_drift` part of the normalization coefficient is wrapped in an absolute value. 

I'm also digging into your approach and seeing if I can crack it with some trig identities. If you don't assume drift velocity along one axis, and instead keep it general, you end up with a fairly compact form for the velocity terms:
`v_r ** 2 - 2 * v_r * v_drift_r * cos(v_phi - v_drift_phi) + v_drift_r **2`

You could then integrate each of the 3 terms separately.

lemmatum replied:
I would like to note, it's astounding how I can't find a single page on the internet that discusses this problem :sob:

dstansby replied:
Yep, I agree that in theory you might be able to integrate those three terms (you only actually need to integrate the middle one, since that's the only one that depends on v_phi. It's possible, and you get a modified Bessel function: https://www.wolframalpha.com/input/?i=integrate+e%5E(-a*cos(x))+between+0+and+2pi

dstansby replied:
So my conclusion is that the function and docstring equation should have a modified Bessel function in it somewhere.

lemmatum replied:
So I'm working this out and am having trouble getting the normalization constant in the case of non-zero drift. That damn Bessel function has a radial velocity and so it shows up in the radial integral from 0 to Inf. [Wolfram doesn't know what to do with it](https://www.wolframalpha.com/input/?i=integrate+x+*+e%5E(-x%5E2+%2F+a%5E2)+*+I_0(b+*+x)+between+0+and+inf)

lemmatum replied:
It looks like the 0th modified Bessel function of the first kind has a [nice series expansion](https://reference.wolfram.com/language/ref/BesselI.html), this would then yield a series of integrals which are just different moments of `v`. Then perhaps there is a way to return the function to some sensible form, maybe another Bessel function? Just some ponderings...

lemmatum replied:
Ooh, I found the first mention of Bessel functions and probability distribution functions! Apparently the p.d.f. [follows a Ricean probability density function](http://elec424.wikia.com/wiki/Modified_Bessel_Functions). Funny how once you find the right jargon you start finding all sorts of [links](https://en.wikipedia.org/wiki/Rice_distribution).

lemmatum replied:
Might be able to use [SciPy's Rice p.d.f. function](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.rice.html).

namurphy commented:
A longer term note...should we perhaps create a `ParticleDistributionFunction` class, with `MaxwellianDistribution`, `KappaDistribution`, and the relativistic equivalent as subclasses?  We could specify the dimension, temperature, particle, and other parameters and then be able to calculate a bunch of different values from each of them.  I think it would be best to merge this pull request before attempting that, so if we think this is a worthwhile idea, then we could create an issue on it.

StanczakDominik commented:
@namurphy that's basically the point of #387 ;)

namurphy commented:
Oh!  Cool!  You're way ahead of me then. :smiley_cat:

lemmatum commented:
So the only thing left is to check whether the normalization coefficient for speed distributions is correct when considering non-zero drift.

StanczakDominik commented on plasmapy/physics/distribution.py:
```diff
@@ -9,14 +9,14 @@
+           "Maxwellian_speed_2D",
+           "Maxwellian_speed_3D",
+           "kappa_velocity_1D",
+           "kappa_velocity_3D"]
```
The reason I like having `]` in its own line and having every entry in a list like this ending with a comma is that when you add an entry, that shows in git as modifying one line instead of modifying 2-3.

lemmatum replied:
Ah, thanks for the tip. I'll keep that in mind for future modifications.

StanczakDominik commented:
That last part sounds like a good unit test!

lemmatum commented:
@StanczakDominik I'm not sure how to unit test that. I think we first need to work through the theory math before designing a unit test. Thoughts?

StanczakDominik commented:
Oh... Yeah, sorry, I skimmed the earlier discussion and thought you had the generic drift velocity case already worked out, just not tested. Yeah, the math comes first ofc.

lemmatum commented:
I'm currently thinking of adding a line to the 2D and 3D speed distributions which raises an exception for non-zero drift velocity and mentions that this feature is a work in progress. Then we can merge this PR and deal with Rician (and whatever the 3D equivalent is) distributions in a separate PR.

namurphy commented:
@lemmatum - Yeah, raising a `NotImplementedError` for the time being and addressing this issue in a separate PR sounds like a good idea.  Again, my thanks to everyone who is working on this!

lemmatum commented:
@namurphy I've added `NotImplementedError` where relevant and modified the pytests for non-zero drift speed. I've also opened issue #462 for follow-up on this. 
Going to merge this PR now.
